Make tables spillable by default

abellina opened the pull request:
Closes https://github.com/NVIDIA/spark-rapids/issues/7672

This depends on https://github.com/rapidsai/cudf/pull/13260 (https://github.com/rapidsai/cudf/issues/13180 overarching issue)

The idea behind this is to be able to add a `Table` as an object into the spillable store. The `Table` will not be made contiguous until that is needed at spill time (to host memory), which is done using `chunked_pack`. Once it has gone through `chunked_pack` the table will be reconstituted as a regular `RapidsDeviceMemoryBuffer` with a single contiguous allocation.

It is draft because I still need to clean it up some and I need to add unit tests specific to this scenario. I also did a bunch of cleanup today and I have not run the latest, so I could have bugs here. But nevertheless, I think it is worth starting to take a look at it.

It won't pass the build because the cuDF PRs are not in yet, and I have one test failure that I know of.

abellina commented on sql-plugin/src/main/scala/com/nvidia/spark/rapids/RapidsBufferCatalog.scala:
```diff
@@ -319,6 +319,50 @@ class RapidsBufferCatalog(
+   * Adds a batch to the device storage. This does NOT take ownership of the
+   * batch, so it is the responsibility of the caller to close it.
+   *
+   * @param batch                batch that will be owned by the store
```
well ok.. it "will" be owned by the store. It isn't "currently" owned as you are adding it, but once the caller closes we definitely own it. So it's definitely in the near future. 

The comment above is to make sure that we don't forget a `withResource` while calling this function.. Some other functions take the object and no `withResource` is needed (e.g. they take ownership).

jbrennan333 replied:
Yeah - I get what you are saying, but the comment on line 326 would make me think I don't need to use `withResource` (although the previous comment clearly indicates need to).  Maybe just say `batch that will be added to the store` for this one so it doesn't sound like a contradiction?

abellina replied:
So for this case will change it to "added". The reason the `Table` takes ownership is because that's the only thing it could do (because `cudf.Table` isn't a regular ref counted cuDF object).

The other add methods are all dealing with objects that are either closeable, or close is a proxy to something else that is (ColumnarBatch is one example of that).

If we make `cudf.Table` ref countable, we can then in `addTable` incRefCount the table itself, then have the caller close it, giving this a more consistent treatment. Does that sounds reasonable? If so I can file a cuDF issue for it.

revans2 commented on tests/src/test/scala/com/nvidia/spark/rapids/RmmSparkRetrySuiteBase.scala:
```diff
@@ -34,6 +34,7 @@ class RmmSparkRetrySuiteBase extends FunSuite with BeforeAndAfterEach {
     }
     val deviceStorage = new RapidsDeviceMemoryStore()
     val catalog = new RapidsBufferCatalog(deviceStorage)
+    RapidsBufferCatalog.setDeviceStorage(deviceStorage)
```
A little confused why this is needed. It looks redundant. Could you at least add in some comments explaining what is happening.

revans2 reviewed (commented):
Mostly nits and comment requests.  But I want to spend some more time looking at the tests.

jbrennan333 commented:
Overall this is excellent work @abellina!   I will likely do another pass, but so far my comments all overlap with those from @revans2.

abellina commented:
Updating the full NDS benchmark with https://github.com/rapidsai/cudf/pull/13260 + https://github.com/rapidsai/cudf/pull/13278 + https://github.com/NVIDIA/spark-rapids/pull/8264 (this), I am seeing a ~4% improvement vs the last nightly. 

I looked at one of the queries (q95) which becomes ~14% faster. Before the change ~7% of kernel time was spent in this function, whereas now it's less than 0.1%.

```
Name = benchmark
Means = 417757.4, 401496.8
Time diff = 16260.600000000035
Speedup = 1.0404999491901306
T-Test (test statistic, p value, df) = 9.948692331384382, 8.820392981431996e-06, 8.0
T-Test Confidence Interval = 12491.56085504279, 20029.63914495728
ALERT: significant change has been detected (p-value < 0.05)
ALERT: improvement in performance has been observed
```

abellina commented on sql-plugin/src/main/scala/com/nvidia/spark/rapids/RapidsHostMemoryStore.scala:
```diff
@@ -62,26 +67,46 @@ class RapidsHostMemoryStore(
+              }
+            } catch {
+              case e: Exception =>
+                hostBuffer.close()
```
changed to closeOnError

abellina commented on sql-plugin/src/main/scala/com/nvidia/spark/rapids/RapidsDeviceMemoryStore.scala:
```diff
@@ -16,48 +16,63 @@
  * Buffer storage using device memory.
- * @param catalog catalog to register this store
+ * @param chunkedPackBounceBufferSize this is the size of the bounce buffer to be used
+ *    during spill in chunked_pack. The parameter defaults to 1MB,
```
I'll sync this with the default in the config

abellina commented:
working on review comments locally and will push but this won't be merge-able until Monday.

abellina commented:
build

abellina commented:
build

abellina commented:
build

abellina commented:
Ok, @revans2 the reason for https://github.com/NVIDIA/spark-rapids/pull/8264/commits/c3c217941b01c00e53ba47f78f616124f3c7fd48 was a change I had made in the test locally and forgotten to undo it. It should be a single close because the code throws a retry and split within a retry (this tests the specific exception), and so every batch that was open was told to close. I do not remember why I had changed this locally, but overall it removes the whole file from the unit test changes.
